cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java

```diff
@@ -113,6 +216,79 @@ public Builder set(String fieldName, @Nullable Object value) {
       return this;
     }
 
+    /**
+     * Sets the date value for {@link Schema.LogicalType#DATE} field
+     *
+     * @param fieldName Name of the field to set
+     * @param localDate Value for the field
+     * @return This builder
+     * @throws UnexpectedFormatException if the field is not in the schema, or the field is not nullable but a null
+     *                                   value is given
+     */
+    public Builder setDate(String fieldName, @Nullable LocalDate localDate) {
+      if (localDate == null) {
+        return set(fieldName, null);
+      }
+      validateAndGetField(fieldName, localDate);
+      fields.put(fieldName, Math.toIntExact(localDate.toEpochDay()));
+      return this;
+    }
+
+    /**
+     * Sets the time value for {@link Schema.LogicalType#TIME_MILLIS} or {@link Schema.LogicalType#TIME_MICROS} field
+     *
+     * @param fieldName Name of the field to set
+     * @param localTime Value for the field
+     * @return This builder
+     * @throws UnexpectedFormatException if the field is not in the schema, or the field is not nullable but a null
+     *                                   value is given
+     */
+    public Builder setTime(String fieldName, @Nullable LocalTime localTime) {
+      if (localTime == null) {
+        return set(fieldName, null);
+      }
+
+      validateAndGetField(fieldName, localTime);
+      Schema fieldSchema = this.schema.getField(fieldName).getSchema();
+      if (fieldSchema.isNullable()) {
+        fieldSchema = fieldSchema.getNonNullable();
+      }
+
+      long nanos = localTime.toNanoOfDay();
+      if (fieldSchema.isLogicalType() && fieldSchema.getLogicalType().equals(Schema.LogicalType.TIME_MILLIS)) {
+        int millis = (int) TimeUnit.NANOSECONDS.toMillis(nanos);
+        fields.put(fieldName, millis);
+        return this;
+      }
+
+      long micros = TimeUnit.NANOSECONDS.toMicros(nanos);
+      fields.put(fieldName, micros);
+      return this;
+    }
+
+    /**
+     * Sets the date value for {@link Schema.LogicalType#TIMESTAMP_MILLIS} or
+     * {@link Schema.LogicalType#TIMESTAMP_MICROS} field
+     *
+     * @param fieldName Name of the field to set
+     * @param zonedDateTime Value for the field
+     * @return This builder
+     * @throws UnexpectedFormatException if the field is not in the schema, or the field is not nullable but a null
+     *                                   value is given
+     */
+    public Builder setTimestamp(String fieldName, @Nullable ZonedDateTime zonedDateTime) {
+      if (zonedDateTime == null) {
+        return set(fieldName, null);
+      }
+
+      validateAndGetField(fieldName, zonedDateTime);
+      Instant instant = zonedDateTime.toInstant();
+      long micros = Math.multiplyExact(instant.getEpochSecond(), 1000000L);
+      long tsMicros = Math.addExact(micros, instant.getNano() / 1000L);
+      fields.put(fieldName, tsMicros);
```

Comment:
this is wrong if logical type is millis

Reply:
This is because we always store `Timestamp` in microseconds in StructuredRecord map. In Time, millis are represented as integer. That is why we can not apply similar logic there.

Reply:
This is because we always set/get timestamp in micros. Even if its millis. Because internally it is always represented as nanos. I will add a comment to clarify it.

Reply:
the physical data we store must match the logical type. If somebody calls the normal .get() method and gets a long back, it must be the timestamp in millis or micros, depending on the the logical type.

Reply:
Ah, yea, I will add that check as well.